task/WG-295-WG-179-WG-185: fix and improve projects with duplicate system path constraints

nathanfranklin opened the pull request:
## Overview: ##

This PR does the following for users:
* allows users to create multiple maps for the same system/path as long as watch_content is False.
* ensure that the user is properly notified (along with https://github.com/TACC-Cloud/hazmapper/pull/255) that they can't create a map syncing the same folder.
* users are now synced every 30 minutes (file syncing still once an hour).

This PR also refactors the code by:
* Dropping the use of ObservableProject and adding a watch_users and watch_content to Project [WG-185](https://tacc-main.atlassian.net/browse/WG-185)
* Improves nginx conf so we see the 409 error in local development.

Related to https://github.com/TACC-Cloud/hazmapper/pull/255


## Related Jira tickets: ##

* [WG-295](https://tacc-main.atlassian.net/browse/WG-295)
* [WG-179](https://tacc-main.atlassian.net/browse/WG-179)
* [WG-185](https://tacc-main.atlassian.net/browse/WG-185)


## Summary of Changes: ##

## Testing Steps: ##
1.  Test with https://github.com/TACC-Cloud/hazmapper/pull/255
2. Try creating multiple maps where `Sync Folder` is not checked.  Having two maps in the same directory should now be allowed
3. Try creating two maps for the same folder and have `Sync Folder` checked.  The creation of the second map should fail.

## Notes: ##

TODO
- [x] separate syncing users and syncing files
- [x] add script to confirm that migration works and everything looks right.
- [x] try locally with production ~and staging~ data to ensure migration is functioning.

taoteg commented on Makefile:
```diff
@@ -10,15 +10,20 @@ help:  ## Display this help screen
-	docker-compose -f devops/docker-compose.local.yml --env-file .env restart workers
+	docker compose -f devops/docker-compose.local.yml --env-file .env restart workers
+
+.PHONY: restart-nginx
```
YAY!!!

taoteg commented on devops/geoapi-services/nginx.conf:
```diff
@@ -38,15 +38,15 @@ http {
+              add_header 'Access-Control-Allow-Origin' '*' always;
+              add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS, HEAD, PUT, DELETE' always;
+              add_header 'Access-Control-Allow-Headers' '*' always;
+              add_header 'Access-Control-Max-Age' 1728000 always;
```
Does this actually override the browser limits enforced in FF, Chrome, etc?

`Maximum number of seconds the results can be cached, as an unsigned non-negative integer. Firefox [caps this at 24 hours](https://searchfox.org/mozilla-central/source/netwerk/protocol/http/nsCORSListenerProxy.cpp#1207) (86400 seconds). Chromium (prior to v76) [caps at 10 minutes](https://source.chromium.org/chromium/chromium/src/+/main:services/network/public/cpp/cors/preflight_result.cc;drc=52002151773d8cd9ffc5f557cd7cc880fddcae3e;l=36) (600 seconds). Chromium (starting in v76) [caps at 2 hours](https://source.chromium.org/chromium/chromium/src/+/main:services/network/public/cpp/cors/preflight_result.cc;drc=49e7c0b4886cac1f3d09dc046bd528c9c811a0fa;l=31) (7200 seconds). The default value is 5 seconds.`

nathanfranklin replied:
I could just remove the `Access-Control-Max-Age` line and go with the default.  thoughts?

nathanfranklin replied:
moving to 86400 on all these.

taoteg commented on devops/local_conf/nginx.conf:
```diff
@@ -13,18 +13,20 @@ http {
+                add_header 'Access-Control-Allow-Origin' '*' always;
+                add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS, HEAD, PUT, DELETE' always;
+                add_header 'Access-Control-Allow-Headers' '*' always;
+                add_header 'Access-Control-Max-Age' 1728000 always;
```
Same question as above for `geoapi-services/nginx.conf`: Does this actually override the browser limits enforced in FF and Chromium?

nathanfranklin replied:
fixed in e4bfad207150aaafcb0f875b914b3b5cd885d8f0

taoteg commented on devops/local_conf/nginx.conf:
```diff
@@ -39,12 +41,13 @@ http {
+                add_header 'Access-Control-Allow-Origin' '*';
+                add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS, HEAD, PUT, DELETE';
+                add_header 'Access-Control-Allow-Headers' '*';
                 add_header 'Access-Control-Max-Age' 1728000;
```
Same question as above (line 25): Does this actually override the browser limits enforced in FF and Chromium?

nathanfranklin replied:
fixed in e4bfad207150aaafcb0f875b914b3b5cd885d8f0

taoteg commented on geoapi/services/projects.py:
```diff
@@ -62,65 +64,32 @@ def create(database_session, data: dict, user: User) -> Project:
+
+                if system_users:
+                    # Initialize the admin status
+                    users_dict = {u.username: u for u in system_users}
```
Being pedantic here, but I feel like this line might read easier as `u for user in system_users`?

nathanfranklin replied:
updated in ca484df0879e4426c3f6e6f3b55adfe3ac6a9ff3

taoteg commented:
We should add logic to the backend routes that prevents a user from creating a new map in a project that is already syncing (even though we do not expose this on the client-side front-end, a user could still call the API and create a map there that violates our syncing constraints).

taoteg commented on geoapi/migrations/versions/6558809eeaed_.py:
Not specific to file, but there is a way to configure alembic so it will generate migration file names that are more human readable:

```
# Using Alembic's Naming Convention:
# Configure Alembic's naming convention in your alembic.ini file:

[alembic]
# ... other configurations
version_locations = migrations/versions
file_template = %%(rev)s_%%(slug)s.py

# Customize the file_template:
# %%(rev)s represents the revision identifier.
# %%(slug)s represents a slugified version of the migration message.
```

Can also reference this example: https://gist.github.com/amanelis/a9fc48039faddcdcbb4a5f2e5fba556a

nathanfranklin replied:
file template approach sounds good.



`file_template = %%(year)d%%(month).2d%%(day).2d_%%(hour).2d%%(minute).2d-%%(rev)s_%%(slug)s`
then `alembic revision --autogenerate -m "add_user_email_column"` produces `20240911_1030-1234abcd_add_user_email_column.py`


if okay with you, I'll add it to the other open https://github.com/TACC-Cloud/geoapi/pull/217/

nathanfranklin replied:
Okay, done in a75f4d8355db95105affec50f1a7d091c9fc9b56

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -375,90 +375,141 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
+
 @app.task()
-def refresh_observable_projects():
+def refresh_watch_content_projects():
```
Pedantic comment #3: the naming of this feels awkward, what do you think about a minor tweak to `refresh_projects_watch_content()`?

nathanfranklin replied:
thanks! will switch.

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -375,90 +375,141 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
 
-    # TODO refactor to consider scaling issues; see https://tacc-main.atlassian.net/browse/WG-47
+@app.task()
+def refresh_watch_users_projects():
```
tic comment #4: What about `refresh_projects_watch_users()`?

nathanfranklin replied:
fixed in e6abc452696186e35253c9b111794b02ef0506ad

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -375,90 +375,141 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
-                            session.commit()
+                    # we need a user with a valid Tapis token for importing files or updating users
+                    importing_user = next(
+                        (u for u in project.users if u.has_unexpired_refresh_token() or u.has_valid_token()), None)
```
Pedantic comment #5: `u for user in project.users` for readability.

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -375,90 +375,141 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
+                        session.add(project)
+                        session.commit()
+
+                        updated_users_to_admin_status = {u.username: u for u in updated_users}
```
Pedantic comment 6: `u for user in updated_users` for readability.

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -375,90 +375,141 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
                     session.rollback()
             total_time = time.time() - start_time
-            logger.info("refresh_observable_projects completed. "
+            logger.info("refresh_watch_users_projects completed. "
```
Pedantic comment 7: `refresh_projects_watch_users()`?

taoteg commented on geoapi/tests/api_tests/test_projects_service.py:
```diff
@@ -46,8 +47,15 @@ def test_create_observable_project(user1,
         'watch_content': True
     }
     proj = ProjectsService.create(db_session, data, user1)
```
Pedantic comment 8: For consistency, wouldn't it be better to use `project` instead of `proj`?

nathanfranklin replied:
i think either proj or project is fine. but updated to `project` in  8af0d2f6f9b6bfb4d6b3719ed5011484773c2978

taoteg commented on geoapi/tests/api_tests/test_projects_service.py:
```diff
@@ -116,11 +120,10 @@ def test_get_features_filter_type(projects_fixture,
         'description': 'new description'
     }
-    proj = ProjectsService.update(db_session, user, projects_fixture.id, data)
+    proj = ProjectsService.update(db_session, projects_fixture.id, data)
```
Pedantic comment 9: For consistency, wouldn't it be better to use project instead of proj?

taoteg replied:
We changed `proj` to `project` across the other files, same here for consistency?

taoteg commented on geoapi/tests/conftest.py:
```diff
@@ -121,33 +120,22 @@ def public_projects_fixture(projects_fixture):
-def observable_projects_fixture():
+def watch_content_users_projects_fixture():
     u1 = db_session.query(User).filter(User.username == "test1").first()
     proj = Project(name="test_observable",
```
Pedantic comment 10: For consistency, wouldn't it be better to use `project` instead of `proj`?

nathanfranklin replied:
updated in f2a4f84325346bc88dc6169f1cdf7c9023b11a5c

taoteg commented on geoapi/tools/check_observable_projects.py:
```diff
@@ -0,0 +1,26 @@
+# Remove this file with https://tacc-main.atlassian.net/browse/WG-377
+
+with create_task_session() as session:
+    observable_projects = session.query(ObservableDataProject).all()
```
This feels off after replacing the refs to `observable...` in other files. Don't really have a great alternative, but maybe something like `watched_projects` to be consistent with the language used in other areas?

nathanfranklin replied:
I think this one is okay. as we'll remove the script once we remove the ObservableDataProject table. I got a follow-on jira issue for it:  https://tacc-main.atlassian.net/browse/WG-377

taoteg commented on geoapi/services/projects.py:
```diff
@@ -352,19 +320,17 @@ def removeUserFromProject(database_session, projectId: int, username: str) -> No
         """
         proj = database_session.query(Project).get(projectId)
```
Pedantic comment 12: For consistency, wouldn't it be better to use `project` instead of `proj`?

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -255,8 +256,8 @@ def import_from_agave(tenant_id: str, userId: int, systemId: str, path: str, pro
 
-    This method is called regularly refresh_observable_projects() and once when a project is created
-    where watch content is true
+    For projects where watch_content is True, this method is called periodically by refresh_watch_content_projects()
```
Pedantic comment 13: would `refresh_projects_watch_content` read a little better?

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -274,7 +275,7 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
     metadata.
 
-    This method is called by refresh_observable_projects() via import_from_agave
+    This method is called by refresh_watch_content_projects() via import_from_agave
```
Pedantic comment 14: would `refresh_projects_watch_content` read a little better?

nathanfranklin replied:
I like it. 💯  updated in e6abc452696186e35253c9b111794b02ef0506ad

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -375,90 +375,141 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
+                                     "Performing rollback of current database transaction")
+                    session.rollback()
+            total_time = time.time() - start_time
+            logger.info("refresh_watch_content_projects completed. "
```
Pedantic comment 15: would `refresh_projects_watch_content` read a little better?

nathanfranklin replied:
good idea.  updated in e6abc452696186e35253c9b111794b02ef0506ad

taoteg reviewed (approved):
Logic all LGTM. I left a slew of comments related to the naming-of-things. Do with them what you wish.

nathanfranklin commented on geoapi/tasks/external_data.py:
```diff
@@ -375,90 +375,141 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
+        Returns None if no such user exists.
+    """
+    importing_user = next(
+        (u for u in project.users if u.has_unexpired_refresh_token() or u.has_valid_token()), None)
```
updated in ca484df0879e4426c3f6e6f3b55adfe3ac6a9ff3

taoteg commented on geoapi/tasks/external_data.py:
```diff
@@ -375,90 +375,142 @@ def import_files_recursively_from_path(session, tenant_id: str, userId: int, sys
-                            session.commit()
+                    # we need a user with a valid Tapis token for importing files or updating users
+                    importing_user = next(
+                        (user for user in project.users
```
Pedantic tweak: `u for user in project.users`

nathanfranklin commented:
> We should add logic to the backend routes that prevents a user from creating a new map in a project that is already syncing (even though we do not expose this on the client-side front-end, a user could still call the API and create a map there that violates our syncing constraints).

Addressed. Turns out we only allow `name`, `description` and `public` to be updated. I added a note to the docstr.

sophia-massie reviewed (approved):
Tested locally and working as expected. LGTM
